marlowe-dashboard-client/src/Template/State.purs

```diff
@@ -0,0 +1,59 @@
+module Template.State
+  ( defaultState
+  , mkInitialState
+  , handleAction
+  ) where
+
+-- Note: There is no independent template state as such (just a property of
+-- the main state). This module simply includes some template-related helper
+-- functions for use in MainFrame.Sate, separated out to keep modules
+-- relatively small and easier to read.
+-- Maybe we could do the same for Contract.State...?
+import Prelude
+import Data.Lens (assign, modifying)
+import Data.Map (Map, insert, fromFoldable)
+import Data.Maybe (fromMaybe)
+import Data.Set (Set)
+import Data.Set (delete, map) as Set
+import Data.Tuple (Tuple(..))
+import Effect.Aff.Class (class MonadAff)
+import Halogen (HalogenM)
+import MainFrame.Lenses (_templateState, _walletState)
+import MainFrame.Types (ChildSlots, Msg)
+import MainFrame.Types (Action, State) as MainFrame
+import Marlowe.Extended (Contract, getParties, getPlaceholderIds, initializeTemplateContent, typeToLens)
+import Marlowe.Semantics (Party(..))
+import Template.Lenses (_contractNickname, _roleWallets, _templateContent)
+import Template.Library (defaultTemplate)
+import Template.Types (Action(..), State, Template)
+
+defaultState :: State
+defaultState = mkInitialState defaultTemplate
+
+mkInitialState :: Template -> State
+mkInitialState template =
+  { template: template
+  , contractNickname: template.name
+  , roleWallets: mkRoleWallets template.extendedContract
+  , templateContent: initializeTemplateContent $ getPlaceholderIds template.extendedContract
+  }
+
+mkRoleWallets :: Contract -> Map String String
+mkRoleWallets contract = fromFoldable $ Set.map (\name -> Tuple name "") (getRoleNames contract)
+
+getRoleNames :: Contract -> Set String
+getRoleNames contract = Set.delete "" $ Set.map roleName (getParties contract)
```

Comment:
This `delete` here removes also `Role ""` which I don't think is right. We can just use `mapMaybe` and have `roleName` return a `Maybe String`. Using `""`, again, is a smell to me, like using `null`

Reply:
Yes, that was bugging me too. Good idea with `mapMaybe`. 👍